# Bibliotecas

In [81]:
from matplotlib import pyplot as plt
import time
import pandas as pd
from typing import Callable
from math import trunc
import numpy as np

# A)

# Método da bisseção

## Explicação

O Método da Bisseção é um algoritmo para encontrar raízes de funções. É preciso um intervalo [a, b], cuja raíz pertence a esse intervalo e que a função seja contínua e que f(a) · f(b) < 0, ou seja, os sinais de f(a) e f(b) sejam diferentes. Durante a execução, o algoritmo irá dividir o intervalo [a, b] em p onde $p = \frac{a + b}{2}$ e seleciona entre [a, p] ou [p, b] o subintervalo onde ainda ocorra a troca de sinais.

<center>
<img src="https://static.wixstatic.com/media/15bb55_f8e8ff43c34944bdb6d680d3ac201503~mv2.png/v1/fill/w_769,h_649,al_c,lg_1,q_90/15bb55_f8e8ff43c34944bdb6d680d3ac201503~mv2.png">
</center>

## Pseudocódigo

0) Seja f(x) contínua em [a, b] e tal que f(a) · f(b) < 0
1) Dados iniciais
    * intervalo inicial [a, b]
    * precisão $\epsilon$

2) Se (b - a) < $\epsilon$, então escolha para $\overline{x}$ qualquer x $\in$ [a, b]. FIM.

3) k = 1

4) M = f(a)

5) $ x = \frac{a+b}{2}$

6) Se M · f(x) > 0, faça a = x. Vá para o passo 8.

7) b = x

8) Se (b - a) < $\epsilon$, escolha para $\overline{x}$ qualquer x $\in$ [a, b]. FIM

9) k = k + 1. Volte para o passo 5.

## Código

In [82]:
def bissecao(f : Callable[[float], float], a : float, b : float, epsilon : float, tempo : bool = False) -> float | tuple[float, float, float]:
    if f(a) * f(b) >= 0:
        raise ValueError("f(a) e f(b) precisam ser de tal forma que f(a) * f(b) < 0")

    if (b - a) < epsilon:
        return (a + b) / 2
    
    tempo_iteracao = []

    while not ((b - a) < epsilon):
        start_time = time.time()

        x = (a + b) / 2

        if f(a) * f(x) > 0:
            a = x
        else:
            b = x

        tempo_iteracao.append(time.time() - start_time)

    if tempo:
        return (a + b) / 2, np.mean(tempo_iteracao)*1000, np.sum(tempo_iteracao)*1000

    return (a + b) / 2

## Exemplo do código funcionando

In [83]:
f = lambda x : x**3 - x - 2 #A raíz é aproximadamente 1.5213797068045
a = 1
b = 2
epsilon = 1e-10
raiz = bissecao(f, a, b, epsilon)
print(raiz)
trunc(raiz * 1e8) / 1e8 == 1.52137970

1.5213797068281565


True

# Método da posição falsa

## Explicação

O método da posição falsa é similar ao método da bisseção, ainda precisando do intervalo [a, b] e que a função seja contínua nesse intervalo. Na execução, também ocorrerá a divisão do intervalo em p, porém agora $p = \frac{a \cdot f(b) - b \cdot f(a)}{f(b) - f(a)}$. Essa equação matém que o ponto p seja escolhido de tal forma que o ponto mais perto da raíz vai puxar o ponto p para próximo de si.

<center>
<img src="https://i.ytimg.com/vi/Htw6P8Yeh0s/maxresdefault.jpg">
</center>

## Pseudocódigo

0) Seja f(x) contínua em [a, b] e tal que f(a) · f(b) < 0

1) Dados iniciais
    * intervalo inicial [a, b]
    * precisões $\epsilon 1 \ \epsilon 2$

2) Se (b - a) < $\epsilon 1$, então escolha para $\overline{x}$ para qualquer x $\in$ [a, b]. FIM.

3) k = 1

4) M = f(a)

5) x = $\frac{a \cdot f(b) - b \cdot f(a)}{f(b) - f(a)}$

6) Se |f(x)| < $\epsilon 2$, escolha $\overline{x}$ = x. FIM

7) Se M · f(x) > 0, faça a = x. Vá para o passo 9.

8) b = x

9) Se (b - a) < $\epsilon 1$, então escolha para $\overline{x}$ para qualquer x $\in$ [a, b]. FIM.

10) k = k + 1. Volte para o passo 5.

## Código

In [84]:
def posicao_falsa(f : Callable[[float], float], a : float, b : float, epsilon1 : float, epsilon2 : float, tempo : bool = False) -> float:
    if f(a) * f(b) >= 0:
        raise ValueError("f(a) e f(b) precisam ser de tal forma que f(a) * f(b) < 0")

    if (b - a) < epsilon1:
        return (a * f(b) - b * f(a)) / (f(b) - f(a))
    
    tempo_iteracao = []
  
    while not ((b - a) < epsilon1):
        start_time = time.time()

        x = (a * f(b) - b * f(a)) / (f(b) - f(a))

        if abs(f(x)) < epsilon2:
            if tempo:
                return x, np.mean(tempo_iteracao)*1000, np.sum(tempo_iteracao)*1000
            return x
        
        if f(a) * f(x) > 0:
            a = x
        else:
            b = x

        tempo_iteracao.append(time.time() - start_time)

    if tempo:
        return (a * f(b) - b * f(a)) / (f(b) - f(a)), np.mean(tempo_iteracao)*1000, np.sum(tempo_iteracao)*1000

    return (a * f(b) - b * f(a)) / (f(b) - f(a))

## Exemplo de código funcionando

In [85]:
f = lambda x : x**3 - x - 2 # A raíz é aproximadamente 1.5213797068045
a = 1
b = 2
epsilon = 1e-10
raiz = posicao_falsa(f, a, b, epsilon, epsilon)
print(raiz)
trunc(raiz * 1e8) / 1e8 == 1.52137970

1.5213797067927124


True

# Método do Ponto Fixo

## Explicação

Dado a função f, escolhe-se uma função $\varphi(x)$ dado que $\varphi(\xi) = \xi$ e assim f($\xi$) = 0. A função $\varphi$ e $\varphi '$ precisam ser contínuas e |$\varphi '$| ≤ M < 1. Dado todas esses condições, escolhe-se um $x_0$, próximo de $\xi$, encontra o $x_1 = \varphi(x_0)$, assim repito todo o processo até chegar em $\varphi(\xi) = \xi$, ou bem próximo 

<center>
<img src="https://blogger.googleusercontent.com/img/b/R29vZ2xl/AVvXsEiS8LKpXmSnrYWkeBW_D6r7ObqoK6-6w-myBH4fJBkxRPBqSq07QXvh03KIdzj59agp6ZHf1CiSd8OVDy3zeYeUwpnn0tdgDD2ffgIFsAbyCkLbR7FPhN4yRpOfldtd5UR7oKMbRNfv_WnU/s1600/pmf01.png">
</center>

## Pseudocódigo

0) Considere a equação f(x) = 0 e a equação equivalente x = $\varphi$(x). Se $\varphi$(x) e $\varphi$'(x) são contínuas em $I$, |$\varphi$'(x)| ≤ M < 1, $\forall$ x $\in$ $I$ e $x_0 \in I $

1) Dados inicias
    * $x_0$: aproximação inicial
    * $\epsilon 1 \ e \ \epsilon 2$: precisões

2) Se |f($x_0$)| < $\epsilon 1$, faça $\overline{x} = x_0 $. FIM.

3) k = 1

4) ${x_1 = \varphi (x_0)}$

5) Se |f($x_1$)| < $\epsilon 1$ ou se |${x_1 - x_0}$| < $\epsilon 2$, faça $\overline{x} = x_1 $. faça $\overline{x} = x_1$. FIM

6) $x_0 = x_1$

7) k = k + 1. Volte para o passo 4.

## Código

In [86]:
def ponto_fixo(g : Callable[[float], float], x0 : float, epsilon1 : float, epsilon2 : float, tempo : bool = False) -> float:
    # Como f(x) = 0 é equivalente a x = g(x), então f(x) = g(x) - x

    f = lambda x : g(x) - x

    if abs(f(x0)) < epsilon1:
        return x0
    
    tempo_iteracao = []

    x1 = g(x0)

    while not (abs(f(x1)) < epsilon1 or abs(x1 - x0) < epsilon2):
        start_time = time.time()

        x0 = x1
        x1 = g(x0)
        
        tempo_iteracao.append(time.time() - start_time)
    
    if tempo:
        return x1, np.mean(tempo_iteracao)*1000, np.sum(tempo_iteracao)*1000
    
    return x1

## Exemplo de código funcionando

In [ ]:
f = lambda x : x**3 - x - 2 # A raíz é aproximadamente 1.5213797068045
g = lambda x : (x + 2)**(1/3) # Escolhemos esse g(x) pois sua derivada é menor que 1 no intervalo [1, 2]
epsilon = 1e-10
x = bissecao(f, 1, 2, 1e-1) # Usamos o método da bisseção para encontrar um x0 próximo da raiz
raiz = ponto_fixo(g, x, epsilon, epsilon)
print(raiz)
trunc(raiz * 1e8) / 1e8 == 1.52137970

1.5213797068424024


True

# Método de Newton-Raphson

## Explicação

O método de newton-raphson é um método que dada uma aproximação inicial $x_0$, encontro f($x_0$) e calculo a reta tangente àquele ponto, depois encontro a raíz dessa reta tangente e chamo de $x_1$. repito o processo até encontrar a raíz da função.

<center>
<img src="https://pages.hmc.edu/ruye/MachineLearning/lectures/figures/NewtonRaphson.png">
</center>

## Pseudocódigo

0) Seja f($\xi$) = 0 e que f(x), f'(x) e f"(x) sejam contínua em $I$ tal que $I$ é um intervalo que contém $\xi$ e f'($\xi$) $\ne$ 0

1) Dados inicias
    * $x_0$: aproximação inicial
    * $\epsilon 1 \ e \ \epsilon 2$: precisões

2) Se |f($x_0$)| < $\epsilon 1$, faça $\overline{x} = x_0$. FIM.

3) k = 1

4) $x_1 = x_0 - \frac{f(x_0)}{f'{x_0}}$

5) Se |f($x_1$) < $\epsilon 1$ ou se |$x_1 - x_0$| < $\epsilon 2$, faça $\overline{x} = x_1$. FIM

6) $x_0 = x_1$

7) k = k + 1. Volte para o passo 4.

## Código

In [88]:
def newton_raphson(f : Callable[[float], float], df : Callable[[float], float], x0 : float, epsilon1 : float, epsilon2 : float, tempo : bool = False) -> float:
    if abs(f(x0)) < epsilon1:
        return x0
    
    tempo_iteracao = []

    x1 = x0 - (f(x0) / df(x0))

    while not (abs(f(x1)) < epsilon1 or abs(x1 - x0) < epsilon2):
        start_time = time.time()

        x0 = x1
        x1 = x0 - (f(x0) / df(x0))
        
        tempo_iteracao.append(time.time() - start_time)

    if tempo:
        return x1, np.mean(tempo_iteracao)*1000, np.sum(tempo_iteracao)*1000

    return x1

# Exemplo de código funcionando

In [ ]:
f = lambda x : x**3 - x - 2 # A raíz é aproximadamente 1.5213797068045
df = lambda x : 3 * x**2 - 1 # Derivada de f e a raíz de f não é a mesma que a raíz de df
epsilon = 1e-10
x = bissecao(f, 1, 2, 1e-1) # Usamos o método da bisseção para encontrar um x0 próximo da raiz
raiz = newton_raphson(f, df, x, epsilon, epsilon)
print(raiz)
trunc(raiz * 1e8) / 1e8 == 1.52137970

1.5213797068045676


True

# Método da secante

## Explicação

O método da secante é similar ao método de newton-raphson, porém em vez de usar uma reta tangente, utiliza-se uma reta secante com dois pontos iniciais, assim encontra a raíz dessa reta e substituio o ponto mais antigo pelo mais novo, e repita até encontrar a raíz

<center>
<img src="https://metodosnumericos.com.br/wp-content/uploads/2020/05/secante-300x196.png" width=500>
</center>

## Pseudocódigo

0) Seja a equação f(x) = 0

1) Dados iniciais
    * $x_0$ e $x_1$: aproximações iniciais
    * $\epsilon 1$ e $\epsilon 2$: precisões

2) Se |f(x0)| < $\epsilon 1$, faça $\overline{x} = x_0$. FIM.

3) Se |f($x_1$)| < $\epsilon 1$ ou |$x_1 - x_0$| < $\epsilon 2$, faça $\overline{x} = x_1$. FIM.

4) k = 1

5) $x_2 = x_1 - \frac{f(x_1)}{f(x_1) - f(x_0)} \cdot (x_1 - x_0)$

6) Se |f($x_2$)| < $\epsilon 1$ ou |$x_2 - x_1$| < $\epsilon 2$, faça $\overline{x} = x_2$. FIM.

7) $x_0 = x_1 \ e \ x_2 = x_1$

8) k = k + 1. Volte ao passo 5.

## Código

In [90]:
def secante(f : Callable[[float], float], x0 : float, x1 : float, epsilon1 : float, epsilon2 : float, tempo : bool = False) -> float:
    if abs(f(x0)) < epsilon1:
        return x0
    
    if abs(f(x1)) < epsilon1 or abs(x1 - x0) < epsilon2:
        return x1
    
    tempo_iteracao = []

    x2 = x1 - (f(x1) * (x1 - x0)) / (f(x1) - f(x0))

    while not (abs(f(x2)) < epsilon1 or abs(x2 - x1) < epsilon2):
        start_time = time.time()

        x0 = x1
        x1 = x2
        x2 = x1 - (f(x1) * (x1 - x0)) / (f(x1) - f(x0))
        
        tempo_iteracao.append(time.time() - start_time)

    if tempo:
        return x2, np.mean(tempo_iteracao)*1000, np.sum(tempo_iteracao)*1000

    return x2

# Exemplo de código funcionando

In [ ]:
f = lambda x : x**3 - x - 2 # A raíz é aproximadamente 1.5213797068045
epsilon = 1e-10
x = bissecao(f, 1, 2, 1e-1) # Usamos o método da bisseção para encontrar um x0 próximo da raiz
x1 = x + 0.5
raiz = secante(f, x, x1, epsilon, epsilon)
print(raiz)
trunc(raiz * 1e8) / 1e8 == 1.52137970

1.5213797068045698


True

# Comparação entre métodos

Primeiro, mostremos uma comparação entre os tempos dada a função $x^3 - x - 2$

In [ ]:
f = lambda x : x**3 - x - 2
g = lambda x : (x + 2)**(1/3)
df = lambda x : 3 * x**2 - 1

bis = bissecao(f, 1, 2, 1e-10, True)
pfalsa = posicao_falsa(f, 1, 2, 1e-10, 1e-10, True)
mpf = ponto_fixo(g, 1.5, 1e-10, 1e-10, True)
mnr = newton_raphson(f, df, 1.5, 1e-10, 1e-10, True)
ms = secante(f, 1.5, 2, 1e-10, 1e-10, True)

dataframe = pd.DataFrame(np.array([bis, pfalsa, mpf, mnr, ms]))
dataframe.columns = ["Raiz Aproximada", "Tempo Médio por Iteração (ms)", "Tempo Total (ms)"]
dataframe.index = ["Bisseção", "Ponto Falso", "Ponto Fixo", "Newton-Raphson", "Secante"]

display(dataframe)

(1.5213797068281565, np.float64(0.0003926894244025735), np.float64(0.0133514404296875))
(1.5213797067927124, np.float64(0.0006399656596936678), np.float64(0.012159347534179688))
(1.5213797067223307, np.float64(0.00010596381293402777), np.float64(0.00095367431640625))
(1.5213797068045751, np.float64(0.00035762786865234375), np.float64(0.0007152557373046875))
(1.5213797068044876, np.float64(0.0004172325134277344), np.float64(0.0016689300537109375))


Raiz Aproximada  Tempo Médio por Iteração (ms)  \
Bisseção                1.52138                       0.000393   
Ponto Falso             1.52138                       0.000640   
Ponto Fixo              1.52138                       0.000106   
Newton-Raphson          1.52138                       0.000358   
Secante                 1.52138                       0.000417   

                Tempo Total (ms)  
Bisseção                0.013351  
Ponto Falso             0.012159  
Ponto Fixo              0.000954  
Newton-Raphson          0.000715  
Secante                 0.001669

Utilizando somente a comparação do tempo, é possível notar que o método da bisseçã e do ponto falso estão duas ordem de grandezas superiores em relação aos outros, porém o tempo não é o único fator importante

Em comparação aos outro métodos, o método da bisseção faz o menor número de operações, o método do ponto falso fica logo atrás. Porém eles são muito demorados para encontrar valores muito precisos, assim sendo ótimos para aproximações inicias, tanto que os outros métodos a utilizam.

O método mais rápido, de maneira geral é o de Newton-Raphson, porém seu maior problema é precisar da derivada da função e às vezes, a derivada não é facilmente encontrada, porém quando se sabe, torna-se a mais rápida. Quando a derivada se torna um empecilho, usa-se o método da secante, que garente uma boa precisão e velocidade, ficando logo abaixo do método de newton.

O método do ponto fixo acaba sendo um método rápido e simples, pois trata apenas de $x_{k+1} = g(x_k)$, porém o maior problema do método é encontrar uma função g(x) que tenha sua derivada menor que 1, pois cado a função g(x) tenha uma derivada maior ou igual a 1, a função diverge.

No fim, a utilização do método depende principalmente da função analisada e da precisão requerida, funções possui uma derivada fácil de encontrar, usa-se newton, quando a derivada é difícil, secante, quando a função é simples e encontrar um g(x) cuja derivada é menor que 1 não é uma tarefa complicada, então usa-se mpf. Por último, caso a precisão não seja muito grande e queira-se uma função que sempre converge para o ponto, usa-se o método da bisseção ou da posição falsa

# B)

# Pseudocódigo para polinômios usando newton-raphson

```py
deltax = x

para k = 1 até itmax, faça
    b = a[n]
    c = b
    
    para i = (n - 1) até 1 faça
        b = a[i] + b * x
        c = b + c * x
    
    b = a[0] + b * x

    Se |b| <= e1 então retorne x
    deltax = b/c
    x = x - deltax
    Se |deltax| <= e2 então retorne x

imprime uma mensagem dizendo que não houve convergência com {itmax} iterações
```

## Pseudocódigo para poliônomios usando método da secante

```py
deltax = x0 - x1

para k = 1 até itmax, faça
    b = a[n]
    c = a[n]
    
    para i = (n - 1) até 1 faça
        b = a[i] + b * x0
        c = a[i] + c * x1
    
    b = a[0] + b * x0
    c = a[0] + c * x1

    deltax = c * deltax / (b - c)

    x2 = x1 - deltax
    
    Se |x2| <= e1 então retorne x2
    x0 = x1
    x1 = x2
    Se |deltax| <= e2 então retorne x2

imprime uma mensagem dizendo que não houve convergência com {itmax} iterações

### Diferenças no pseudocódigo

A primeira diferença foi no deltax. O deltax no pseudocódigo anterior representa a diferença entre $x_k - x_{k-1}$. Agora ele reprenta $x_{k+1} - x{k}$, por isso a formula precisou ser alterada. quando fazemos $c \cdot deltax \div (b - c)$ estamos fazendo $f(x_1) \cdot (x_0 - x_1) \div (f(x_0) - f(x_1)) = f(x_1) \cdot (x_1 - x_0) \div (f(x_1) - f(x_0))$. Sendo essa a principal mudança do codigo.

Outras mudanças são que b e c significam respectivamente f($x_0$) e f($x_1$), além de toda a difereça dos dois algoritmos como a fato que o método da secante precisa de dois pontos já o método de newton precisa só de 1

## Pseudocódigo para polinômios usando método do ponto fixo

```py
deltax = x

para k = 1 até itmax, faça
    b = a[n]
    
    para i = (n - 1) até 1 faça
        b = a[i] + b * x
    
    g = -1 * a[0] / b
    b = a[0] + b * x

    Se |b| <= e1 então retorne x
    deltax = g - x
    x = g
    Se |deltax| <= e2 então retorne x

imprime uma mensagem dizendo que não houve convergência com {itmax} iterações
```

### Diferenças no pseudocódigo

A maior diferença se refere-se ao g. Todo o polinômio P(x) pode ser escrito como $a_0 + x(a_1 + x(...))$, assim caso queremos uma função $\varphi$, onde $\varphi(\xi) = \xi$, podemos isolar o x mais externo de P(x), assim $x = -a_0 \div (a_1 + x(...))$, isso garante que a função não exploda e mantém sua derivaa menor que um, garantindo todas as características fo MPF. Utilizando as nomenclatura dos polinômios em função de b, fica que $\varphi(x) = -a_0 \div b_1$

De resto é o mesmo, calcula P($x_0$) e $\varphi(x_0)$, se P($x_0$) for suficientemente baixo pare, senão altere x para g, se a diferença entre x para g é suficientemente baixa, pare, senão continue até parar

# Métodos adaptados

In [201]:
def newton_pol(a : list[float], x : float, itmax : int, epsilon1 : float, epsilon2 : float) -> float:
    deltax = x
    tempo_iteracao = []

    for k in range(itmax):
        start_time = time.time()

        b = a[-1]
        c = b

        for i in range(len(a) - 2, 0, -1):
            b = a[i] + b * x
            c = b + c * x
        
        b = a[0] + b * x
        deltax = b / c

        if abs(b) <= epsilon1:
            tempo_iteracao.append(time.time() - start_time)
            return x, np.mean(tempo_iteracao)*10e6, np.sum(tempo_iteracao)*10e6, k, abs(deltax)
        
        x = x - deltax

        tempo_iteracao.append(time.time() - start_time)

        if abs(deltax) <= epsilon2:
            return x, np.mean(tempo_iteracao)*10e6, np.sum(tempo_iteracao)*10e6, k, abs(deltax)
        
    print("Número máximo de iterações atingido e o algorimo não convergiu.")

def secante_pol(a : list[float], x0 : float, x1 : float, itmax : int, epsilon1 : float, epsilon2 : float):
    deltax = x1 - x0
    tempo_iteracao = []

    for k in range(itmax):
        start_time = time.time()

        b = a[-1]
        c = a[-1]

        for i in range(len(a) - 2, -1, -1):
            b = a[i] + b * x0
            c = a[i] + c * x1

        deltax = c * deltax / (b - c)
        x2 = x1 - deltax

        if abs(c) <= epsilon1:
            tempo_iteracao.append(time.time() - start_time)
            return x2, np.mean(tempo_iteracao)*10e6, np.sum(tempo_iteracao)*10e6, k, abs(deltax)
        
        x0 = x1
        x1 = x2

        tempo_iteracao.append(time.time() - start_time)

        if abs(deltax) <= epsilon2:
            return x2, np.mean(tempo_iteracao)*10e6, np.sum(tempo_iteracao)*10e6, k, abs(deltax)

    print("Número máximo de iterações atingido e o algorimo não convergiu.")

def mpf_pol(a : list[float], x: float, itmax : int, epsilon1 : float, epsilon2 : float):
    deltax = x
    tempo_iteracao = []

    for k in range(itmax):
        start_time = time.time()

        b = a[-1]

        for i in range(len(a) - 2, 0, -1):
            b = a[i] + b * x
        
        try:
            g = -1 * a[0] / b
        except:
            g = -1 * a[0] / 10e-30
        b = a[0] + b * x

        deltax = g - x

        if abs(b) <= epsilon1:
            tempo_iteracao.append(time.time() - start_time)
            return x, np.mean(tempo_iteracao)*10e6, np.sum(tempo_iteracao)*10e6, k, abs(deltax)
        
        x = g

        tempo_iteracao.append(time.time() - start_time)

        if abs(deltax) <= epsilon2:
            return x, np.mean(tempo_iteracao)*10e6, np.sum(tempo_iteracao)*10e6, k, abs(deltax)

    print("Número máximo de iterações atingido e o algorimo não convergiu.")

In [217]:
a = [-6.8, 10.8, -10.8, 7.4, -3.7, 1]  # Coeficientes do polinômio começando pelo a0 e terminando no an
epsilon = 10e-8
itmax = 100
resultados = [newton_pol(a, 1.5, itmax, epsilon, epsilon), secante_pol(a, 1, 2, itmax, epsilon, epsilon), mpf_pol(a, 1.5, itmax, epsilon, epsilon)]
df = pd.DataFrame(resultados)
df.columns = ["x aproximado", "Tempo médio por iteração (µs)", "Tempo total (µs)", "Número de iterações", "Passo final"]
df.index = ["Método de newton", "Método da secante", "Método do ponto fixo"]
display(df)

print(f"Método de newton: {resultados[0][0]}")
print(f"Método da secante: {resultados[1][0]}")
print(f"Método do ponto fixo: {resultados[2][0]}")

x aproximado  Tempo médio por iteração (µs)  \
Método de newton               1.7                       8.106232   
Método da secante              1.7                       6.622738   
Método do ponto fixo           1.7                       3.511841   

                      Tempo total (µs)  Número de iterações   Passo final  
Método de newton             40.531158                    4  5.744552e-08  
Método da secante            59.604645                    8  1.150482e-08  
Método do ponto fixo        259.876251                   73  9.718606e-08

Método de newton: 1.7000000000000055
Método da secante: 1.7000000000001987
Método do ponto fixo: 1.6999999561681662


Perceba que todos chegaram ao valor aproximado de 1.7, mas com diferenças. O tempo médio de cada iteração é maior no método de newton que no MPF, porém o tempo total do MPF é maior, isso ocorre pois os cálculos que o método de newton executa são mais complexos que o MPF, mas o número de repetições que o MPF precisa executar é enorme, assim provando que o método de newton é melhor para polinômios, pois é fácil encontrar suas derivadas